libraries/botframework-connector/src/auth/microsoftAppCredentials.ts

```diff
@@ -1,83 +1,156 @@
+        ['state.botframework.com', new Date(8640000000000000)]              // Date.MAX_VALUE
+    ]);
 
-    private static refreshEndpoint: string = AuthSettings.refreshEndpoint;
-    private static refreshScope: string = AuthSettings.refreshScope;
+    private static readonly cache: Map<string, OAuthResponse> = new Map<string, OAuthResponse>();
 
-    public constructor(credentials: BotCredentials) {
-        if (typeof credentials !== 'undefined') {
-            if (typeof credentials.appId !== 'undefined') {
-                this.appId = credentials.appId;
-            }
+    appPassword: string;
+    appId: string;
 
-            if (typeof credentials.appPassword !== 'undefined') {
-                this.appPassword = credentials.appPassword;
-            }
-        }
+    readonly oAuthEndpoint: string = Constants.ToChannelFromBotLoginUrl;
+    readonly oAuthScope: string = Constants.ToChannelFromBotOAuthScope;
+    readonly tokenCacheKey: string;
+
+    constructor(appId: string, appPassword: string) {
+        this.appId = appId;
+        this.appPassword = appPassword;
+        this.tokenCacheKey = `${appId}-cache`;
     }
 
-    public signRequest(webResource: msrest.WebResource): Promise<msrest.WebResource> {
-        return new Promise((resolve, reject) => {
-            if (this.appId !== '' && this.appPassword !== '') {
-                this.getAccessToken((err, token) => {
-                    if (!err && token) {
-                        var tokenCredentials = new msrest.TokenCredentials(token);
-                        tokenCredentials.signRequest(webResource).then(resolve, reject);
-                    } else {
-                        reject(err);
-                    }
-                });
+    async signRequest(webResource: msrest.WebResource): Promise<msrest.WebResource> {
+        if (this.shouldSetToken(webResource)) {
+            let token = await this.getToken();
+            return new msrest.TokenCredentials(token).signRequest(webResource);
+        }
 
-            } else {
-                resolve(webResource);
+        return webResource;
+    }
+
+    async getToken(forceRefresh: boolean = false): Promise<string> {
+        if (!forceRefresh) {
+            // check the global cache for the token. If we have it, and it's valid, we're done.
+            let oAuthToken = MicrosoftAppCredentials.cache.get(this.tokenCacheKey);
+            if (oAuthToken) {
+                // we have the token. Is it valid?
+                if (oAuthToken.expiration_time > new Date(Date.now())) {
+                    return oAuthToken.access_token;
+                }
             }
-        });
+        }
+
+        // We need to refresh the token, because:
+        // 1. The user requested it via the forceRefresh parameter
+        // 2. We have it, but it's expired
+        // 3. We don't have it in the cache.
+        let oAuthToken = await this.refreshToken();
```

Comment:
>let oAuthToken = await this.refreshToken(); [](start = 8, length = 43)

We recently made a change to the v3 code to ensure that we only attempt 1 refresh of the token at a time to prevent the login service from timing out.  The way this is currently coded if we were to get hammered with 100 simultaneous messages on startup we would attempt to retrieve 100 simultaneous tokens which will cause timeouts.  We need to wrap the refresh() call with a promise that everyone joins on to wait for the refresh to complete. You can review my v3 code to see how this works.

Reply:
fixed